### ISSUE TITLE
Feature: Provide way for user to not get VT sequences in output

### ISSUE THREAD

**Issue by SteveL-MSFT** [PowerShell/PowerShell#10811]
# Summary of the new feature/enhancement

Update Out-File to strip VT by default, add `-IncludeEscapeSequences` type switch if user wants it in output
Add `$OutputPreference` variable where the user can set it globally
Update internal LineOutput to strip VT based on `$OutputPreference`

**Comment by rkeithhill** [comment]
Should this be encoded into each `Out-*` cmdlet (actually thinking of set-clipboard here)?  Maybe there should be a single, intermediate cmdlet to strip VT seqs?

**Comment by SteveL-MSFT** [comment]
@rkeithhill file redirection and clipboard are the two scenarios I'm thinking of

**Comment by jhoneill** [comment]
What's the solution when piping to non PS commands (e.g. | clip). 

Really.
- Text output should be kept **clean.**  That's been a basic tenet of PowerShell from the start. 
- Hard-coding things to appear in the writer's favorite color should be strongly discouraged. The chosen color may be difficult to read for some people or in some configurations. (Treating the PSReadline colors  or $host.privateData as a "theme" is a better approach, these will likely have been customized to something which will work)   
-  It is possible to tolerate prettification and "reverse engineer" back to clean text. However re-writing _individual_ PS commands to remove authors' decoration is a bad way to do it - all non PS commands would get "dirty" text, etc. 

In some places color helps. (e.g. PSReadline color coding syntax; the parser error messages in the current preview) and in these cases the text is only going into the console to be consumed by human eyeball. If it goes anywhere else, mixing non-output (like VT sequences) with "actual" output and sending both to std-out is the path to madness.

**Comment by mklement0** [comment]
Agreed, @jhoneill.

We should consider taking a page out of the book of Unix utilities that provide colored output, such as (GNU) `grep`:

* Do _not_ use VT sequences, _except_ when a cmdlet's "stdout" (success output stream) is connected to a _terminal_ (implicit output formatting) or a _formatting cmdlet_, 
   * That would require special-casing for `Out-File` and `Out-String`, however, given that they use the formatting system too.

  * In fact, arguably, coloring should be confined to the _formatting layer_ - though that's probably not feasible in cases where selective coloring requires knowledge of the details of the output objects, such as in the case of `Select-String`.
  
  * Therefore, the onus is on individual cmdlets that potentially provide VT sequences as part of their _data_ to suppress them when not (ultimately) printing to the screen (by default) and to respect a preference variable that controls the behavior (see next point).

* Provide _categorical suppression of_ and _unconditional_  use of VT sequences via an _opt-in_ mechanism, as a preference variable and possibly a new common parameter (though that may be overkill, given that few cmdlets produce colored output).

  * I also suggest naming the preference variable something like `$ColoringPreference`, with enumeration values `HostOnly` (default), `Always`, `Never`.

In other words: _By default_ you should only ever get VT sequences if you _print to the screen_.

**Comment by SteveL-MSFT** [comment]
@jhoneill just to be clear, the colors are completely user configurable.

@mklement0 limiting VT to screen output only makes sense, need to figure out how easy or hard it would be to do that

**Comment by jhoneill** [comment]
@mklement0  :-)  Select-string - or to be more correct, the matchinfo object - is something of a basket case, doing the formatting inside the ToString() method of the object, based on a private property set by select-string. That fails if the object is serialized (e.g. in remoting)  I put together a .format.ps1xml file for match info, which is the proper mechanism for this stuff ... 

![image](https://user-images.githubusercontent.com/16209593/67039042-7d95f700-f118-11e9-9c22-a03fc473859f.png)

@SteveL-MSFT  when you say _completely_ could I change the colors for the labels in `get-error` to be magenta instead of green ? 
 That bit of format-xml, shows that signalling that color is OK might lead to some nasty things being hard coded (please, no-one say you _like_ those colors !).  _But_ the script block in the formatting XML _could_ heed a preference variable and/or get values from "themes". AFAIK you can't use a script block for a label in the formatting XML so that route to colored labels won't work. 
But _Labels_ in tables and lists should not change color from one command to another (one preference variable which applied to all of them ... perhaps).     


BTW did you know that` get-formatData `doesn't work if the data was compiled, it only shows if data comes from a free-standing XML file (about_format.ps1xml still refers to the files which used to hold the bulk of the data)  and `update-formatData` updates the data part but not the labels - shall I open issues for those ?

**Comment by mklement0** [comment]
@jhoneill: Nice - confining the coloring to the formatting layer is definitely the right approach; does it perform acceptably?

As for `Get-FormatData`: the compiled formatting data does show in principle, but there's a bug / quirk with respect to certain types - see #4237; tl;dr: use `get-formatData -PowerShellVersion $PSVersionTable.PSVersion` to see everything.

**Comment by lzybkr** [comment]
@SteveL-MSFT - programs use `GetConsoleMode` on Windows or `isatty` on Linux to determine if stdout/stderr is going to a screen or not.

**Comment by SteveL-MSFT** [comment]
@jhoneill see `$Host.PrivateData` on colors that can be configured.  You can make it Magenta if you'd like.

**Comment by jhoneill** [comment]
@SteveL-MSFT  FormatAccentColor is new.  
Using $host.private data / $host.ui.foregroundColor / $host.ui.foregroundColor / psreadline colors side-steps the problem of the user not getting on with the selected color. [Docs should explain hard coded colors: bad , "theme" colors: better] 
Then it's just a case of dealing with escape sequences. 

BTW if you set foreground color via $host and not in on the shell properties; then  if something sends "`e[0" it resets things to whatever the terminal says it should be and your setting is lost. I wondered if I could use  this for mischief ... 
![image](https://user-images.githubusercontent.com/16209593/67061191-7215f080-f157-11e9-8cf3-6677d2465c63.png)

So I've embedded an escape code in the user's description. I output the description and my foreground colour changes. Then it gets weird. Anything that formats the object as a table causes the terminal to revert to default colours as soon as that field is output. It didn't _change_ the colour because it was already changed, but the escape sequence cause the colour to change from the next column . If I use format-list it does what I'd expect, the colour changes from when the description is output and stays changed. 
![image](https://user-images.githubusercontent.com/16209593/67061535-86a6b880-f158-11e9-9216-2746ea93a8e8.png)
 




@mklement0  Perf is not visibly different. And between daily builds tostring on matchinfo has gone back to normal and following your tip I can see it is calling  toEmphasizedString instead.

**Comment by SteveL-MSFT** [comment]
@lzybkr actually, that won't help in this case as within the same pwsh interactive session, I believe the desire is to have colors when we know the output isn't being redirected or piped.  I believe scrubbing it in  Host APIs is the way to go.

@jhoneill the reset code was added to fix another problem where a user would deliberately insert a VT code for their own coloring but when formatted as a table and is truncated, then it would bleed into other columns. 

I don't think it makes sense to invent a new markup in PS for VT codes as VT esc seq is that markup.  Coloring should be in formatting which it currently primarily is with the exception of `select-string` as pointed out (even this could be moved to formatting if someone wanted to spend the time to do it).  The concept of stripping VT codes is not optimal, but seems like the best approach to minimize complexity in formatters and the fact that they can't know where the output is going as that's handled much lower in the engine.

**Comment by jhoneill** [comment]
**Additional** (@mklement0  re format xml). I checked what happens on Server 2012 R2 as that is in extended support until October 2023 , and its console host doesn't support escape sequences.  [Side note  I remember ANSI.SYS in DOS in c. 1989, but NT 3.1 to Windows 8.1 seemed to manage without it. It's weird having it back and it doesn't feel like an old friend]
On 2012 R2, the new formatting for parser errors, get-error, and select-string are all smart enough _not_ to send escape sequences to the console. My format XML isn't ; char 27 prints as a backward arrow, and the rest of the sequence prints to the screen. 

Generically, code in a format.ps1XML needs to check  `$Host.ui.supportsVirtualTerminal `  & a preference variable, and select a "theme" color (not a hard coded author favorite) .  Really there should be some helper code for that, so you write `render($_.property,VerboseForeground)` or `render('Label',ReadLineEmphasis)`  and it should NOT send a "reset terminal to default" at the end but should check` $host.ui.rawui.data ` and set the foreground / background back to to those values (and turn off inverse or underline). 
For Select-String to support -No emphasis _and_ work with formatting XML the Emphasis property for matchInfo objects needs to be public, and then the format-XML code can check it...  

I think I can summarize my view after thinking way too much about it as  :

1. Data is colorless. Unless what is being written targets eyeballs only, and the color makes things clearer (not nicer to to look at), don't do it.  The new parser error message is a case where color does make things clearer and (as an error which won't be caught and redirected) it's eyeballs only. That's the right use. Get-Error, if it needs color at all, uses it because the indents are ragged and the labels blend in with the values. The solution to this is to fix the layout, not color the labels. It's less "eyeballs only" we might ask someone to run get-error and put the results into a problem ticket... so this looks like the wrong use. Select-string benefits from the emphasis, especially when seeking a complicated regex. But it might be sent somewhere hence a need to manage it.       
2. Where color is beneficial but adding it has downstream effects [it didn't _target_ downstream use, but users did that anyway], it is better to stop it at source than to strip it later. 
3. Using format data is better than hard coding formatting in cmdlets and their objects.
4. What goes out as a PowerShell release will set a standard; so it needs to demonstrate good practice....  I think for something which "does" color **to make it into a release it should be  meaningful, optional, well behaved and consistently implemented**

**Comment by jhoneill** [comment]
> @jhoneill the reset code was added to fix another problem where a user would deliberately insert a VT code for their own coloring but when formatted as a table and is truncated, then it would bleed into other columns.

One should probably do the same format-list ... but also it should reset to the foregroundColor in $host.ui not esc[0m which resets to the terminal default. 

> I don't think it makes sense to invent a new markup in PS for VT codes as VT esc seq is that markup. 
Even if you invented a new mark up, VT codes are still there, so people will still use them and you'll have two forms.   
>Coloring should be in formatting which it currently primarily is with the exception of `select-string` as pointed out (even this could be moved to formatting if someone wanted to spend the time to do it). 
Agree ! I feel like I know enough to do it now :-) 
>The concept of stripping VT codes is not optimal, but seems like the best approach to minimize complexity in formatters and the fact that they can't know where the output is going as that's handled much lower in the engine.

See the comment above:  if it is done in the formatting and there is a standard "format this" which can does nothing if $colorPreference is "None" or the terminal doesn't support it;  sets colors, reverse field and underline, and reverts back to the the right colors instead of doing a terminal reset that would be good.  I might prototype that first ...

**Comment by lzybkr** [comment]
@SteveL-MSFT - Checking isatty/GetConsoleMode is necessary when PowerShell is invoked from another process, e.g. ADO or in a pipeline from a parent shell (bash/PowerShell/etc) process.

Most CLIs also offer an option to override the isatty check and force colors. I know for certain ADO would like such an option so they can display the colors in CI builds in the same way that you see on the screen.

**Comment by SteveL-MSFT** [comment]
After some discussion with @JamesWTruher who has experience in this space, we're going to defer this to the next release.  What we really need is a platform solution where cmdlets/formatters can provide rendering attributes to text then the render can decide what to do with it.  In cases where text is passed in the pipeline, it would just be the text as the rendering attributes are a separate property.  Only renders that understand those attributes would render them as VT escape sequences (or other renderings as necessary).  For the developer experience, we discussed a StringBuilder type experience where in addition to Append(), you could Append(attribute) where attribute could be color, bold, underline, etc...  More thinking is needed on this which is why it won't make it into PS7, but something we can target for next release.

**Comment by jhoneill** [comment]
@SteveL-MSFT  I've just had an idea ... all the cases which I know of which use escape sequences check `$host.ui.SupportsVirtualTerminal`

So if that class had a method `DisableVirtualTerminal()` which caused supportsVT return false for the rest of the PowerShell session (and possibly .ResetVirtualTerminal which checked if it was supported and if so turned it back on)  that would meet the objective. Doesn't need a preference variable, can go in a script or profile ... not very hard  to implement...

**Comment by SteveL-MSFT** [comment]
@jhoneill implementation isn't really the concern but how to expose to users and how to expose to developers.  Exposing a method to change the value of `SupportsVirtualTerminal` is easy, but not discoverable for users.  Also, the general consensus is that users probably want VT for the interactive shell and output to the host (PSReadLine depends on some of this, for example), just not in the output going through the pipeline or redirected.

**Comment by jhoneill** [comment]
The plus side is if developers check for` SupportsVirtualTerminal `(or use something which checks for them) they is nothing _they_ need to do. 
I think you're right it's only a few users who want _all_ VT off , _all_ the time. Toggling it  either side of a command so you can send non-formatted stuff somewhere, but retain syntax highlighting (for example) doesn't seem an outrageous overhead. You could have Reset-Color and Disable-Color cmdlets to wrap the changes to ` SupportsVirtualTerminal `  for discoverablity . 

An alternative is to put a "stripVT" switch in Out-string to do a regex replace on "\e\[\d.*?m" , but not sending is always better than stripping out. 

I understand the desire to have a really top solution for the next release. However, with Show-Markdown, Get-error and Select-String all sending VT and writers seeing it as an easy way to make stuff pretty , I don't think the need to stop VT going to the output stream will go away. The sooner those writers can be steered towards a good practice, the smaller the number of horrors that will need to be cleaned up.

**Comment by SteveL-MSFT** [comment]
I expect we'll have this address in the release after PS7.  The current concept is to have separation of the text and attributes.  This makes it easy to pass both and only formatting (ideally) would use the attributes to do the rendering.  This also makes it transparent to everything else already in PowerShell as `ToString()` (for example) would just be the naked string sans VT.  The other benefit of separating the attributes is that the color palette can be customized.  So for some customers, they may want underlined text to be bold and underlined to make it easier to read.  This can all be customized.

I also want it to be easier for developers writing new scripts/cmdlets to participate.  So instead of understanding and hardcoding VT codes, they can get an experience more like (note, class names are placeholders):

```powershell
$s = [PSStringBuilder]::new()
$s.AppendAttribute('Red')
$s.Append("This is in red,")
$s.AppendAttribute('Inverse','Bright','Green')
$s.Append("this is not")
$s.AppendAttribute('Reset')
$s.ToString()

This is in red, this is not

$s.RenderString()

.... well, you get the idea :)
```

**Comment by SteveL-MSFT** [comment]
@lzybkr had a good suggestion to have a param on `pwsh` like `-color` with `always, auto, never` values.  Cmdlets can still use an api like above to declare attributes and the formatter can deal with whether to generate the VT or not.

**Comment by jhoneill** [comment]
> @lzybkr had a good suggestion to have a param on `pwsh` like `-color` with `always, auto, never` values. Cmdlets can still use an api like above to declare attributes and the formatter can deal with whether to generate the VT or not.

It is a good suggestion, it will solve the problem for a good proportion of cases. 
My instinct says its a only small minority of cases where you'd be in a session and see you needed to turn off VT and starting a new session wouldn't work, but I don't know how to gauge it.

**Comment by DHowett-MSFT** [comment]
Just wanted to jump in and caution against this:

> should reset to the foregroundColor in $host.ui 

`ForegroundColor`, barring some new version of .NET’s Console class that supports terminals that aren’t Windows well, will never be the right thing to use here. Terminals do not, broadly speaking, have an application-knowable foreground or background color in the general sense. Instead of saving/restoring the foreground color index, a compliant application should emit `\e[39m` to tell the terminal to restore to _its_ default foreground color.

Every time a commandline application thinks it knows better than the terminal it’s connected to what the user wants on the screen, it is ill-formed. :smile:

**Comment by jhoneill** [comment]
> Just wanted to jump in and caution against this:
> 
> > should reset to the foregroundColor in $host.ui
> 
> `ForegroundColor`, barring some new version of .NET’s Console class that supports terminals that aren’t Windows well, will never be the right thing to use here. Terminals do not, broadly speaking, have an application-knowable foreground or background color in the general sense. Instead of saving/restoring the foreground color index, a compliant application should emit `\e[39m` to tell the terminal to restore to _its_ default foreground color.
> 
> Every time a commandline application thinks it knows better than the terminal it’s connected to what the user wants on the screen, it is ill-formed. 😄


**If** a user sets foreground / background color  via $host.ui  and select-string etc. changes it to terminal defaults that's not well informed either. 
For example 


@lzybkr and I discussed this on https://github.com/PowerShell/PSReadLine/issues/1110 there is another reason for not setting the color back, and that is that the terminal can use a custom setting which doesn't match one of the 16 standard colors, so there is a choice either set the color which foreground color says it should be and if a custom color has been set you don't honor it. Or force everything bea

**Comment by heaths** [comment]
Re: https://github.com/PowerShell/PowerShell/issues/10811#issuecomment-545153262

A good idea to make it easy, but if this ends up being the route, allow use of VT codes for more advanced scenarios where we want full access to the 256 color table and other modifiers.

**Comment by SteveL-MSFT** [comment]
.NET is building a new System.CommandLine library.  One of the areas they are covering is this problem of strings and decoration.  So for now, we'll work with them and adopt their solution.

**Comment by SteveL-MSFT** [comment]
I played with System.CommandLine.Rendering library and it's suitable for PowerShell needs.  Here's an example in script:

```powershell
using namespace System.CommandLine.Rendering

add-type -AssemblyName ./System.CommandLine.Rendering.dll

$cs = [ContainerSpan]::new(
    [StyleSpan]::UnderlinedOn(),
    [ForegroundColorSpan]::Green(),
    [BackgroundColorSpan]::Yellow(),
    [ContentSpan]::new("hello"),
    [StyleSpan]::UnderlinedOff()
    )

$cs += [ForegroundColorSpan]::Rgb(225, 110, 132)
$cs += [BackgroundColorSpan]::Reset()
$cs += [ContentSpan]::new(" there")

$sw = [System.IO.StringWriter]::new()
$cs.WriteTo($sw, [OutputMode]::Ansi)
$sw.ToString()
```

Integrating this into PowerShell is pretty straightforward except for integrating with existing color maps.

Currently, a PSHost can expose some colors to use:

```powershell
PS> $Host.PrivateData
FormatAccentColor       : Green
ErrorAccentColor        : Cyan
ErrorForegroundColor    : Red
ErrorBackgroundColor    : -1
WarningForegroundColor  : Yellow
WarningBackgroundColor  : -1
DebugForegroundColor    : Yellow
DebugBackgroundColor    : -1
VerboseForegroundColor  : Yellow
VerboseBackgroundColor  : -1
ProgressForegroundColor : Black
ProgressBackgroundColor : Yellow
```

These are all of type `[System.ConsoleColor]` which isn't used by `System.Commandline.Rendering`.  Whereas `$Host` has separate foreground and background colors, those are now distinct types `ForegroundColorSpan` and `BackgroundColorSpan`.  These existing members will be retained for compatibility reasons (cmdlets or scripts may query for those values expecting them to exist and be populated), but would not be used by PowerShell formatting.

## Proposal

Add new member to `PrivateData` called `Style`.  `Style` would contain properties of type `ContentSpan` for

- FormatAccentStyle = [ForegroundColorSpan]::Green()
- ErrorAccentStyle = [ForegroundColorSpan]::Cyan()
- ErrorStyle = [ForegroundColorSpan]::Red()
- WarningStyle = [ForegroundColorSpan]::Yellow()
- DebugStyle = [ForegroundColorSpan]::Yellow()
- VerboseStyle = [ForegroundColorSpan]::Yellow()
- ProgressStyle = [ForegroundColorSpan]::Black() + [BackgroundColorSpan]::Yellow()

## Selective output of ANSI

Cases where`[OutputMode]::Plaintext` is used:

- When redirecting output
- When piping output to a native command

otherwise, the value of `$PSOutputMode` is used

## Preference variable

A `$PSOutputMode` variable will allow custom control with default value of `[OutputMode]::Auto`

**Comment by JamesWTruher** [comment]
@SteveL-MSFT - While this might be excellent for providing a way to easily attribute strings, I'm not sure how this helps the prime motivation. How do we know whether we should emit the escape sequence or not. The example for creating the escape sequences is done _a priori_ - it's the `WriteTo` method as `ansi` which has decided to add the attributes. How would this work in PowerShell?  Also, the logic for table creation in PS is done against the content of the field. Would the field include those ansi sequences? This would cause difficulties when calculating the column width. If not, how does the association of attributes survive the table creation logic?

**Comment by SteveL-MSFT** [comment]
@JamesWTruher In our code, we would `WriteTo()` a string, but setting the `OutputMode` parameter dictates if it'll include the ANSI escape sequences or not.

**Comment by jhoneill** [comment]
I must have a play with the sample, but this sounds _exactly_ what I've been asking for. 
The one challenge is that code written to use it won't be compatible with older PS versions, but that's unavoidable.

**Comment by felixfbecker** [comment]
Stronly agree with the point that all coloring should happen in the _formatting_ system, i.e. in Format.ps1xml files, _not_ by changing the _data_ to include VT sequences.

This is how I implemented it for multiple modules:

- https://github.com/pcgeek86/PSGitHub#pretty-formatting
- https://github.com/felixfbecker/PowerGit#how-to-use
- https://github.com/sourcegraph/PSSourcegraph#sourcegraph-for-powershell

**Comment by felixfbecker** [comment]
Also I'd LOVE to have `System.CommandLine.Rendering` always available in PowerShell and being able to use it in Format.ps1xml files to ensure control sequences are always properly terminated etc. It would be nice to have some of the API also available through declarative XML in addition to at runtime through .NET.

One thing I'm a bit confused by is that the API is 
```powershell
[StyleSpan]::UnderlinedOn()
"text"
[StyleSpan]::UnderlinedOff()
```
as opposed to
```powershell
[StyleSpan]::Underlined("text")
```

which is how e.g. [`chalk`](https://github.com/chalk/chalk) works. The latter API seems a lot less foot-gunny as you cannot forget to turn underline off again.

**Comment by SteveL-MSFT** [comment]
@felixfbecker the `System.CommandLine.Rendering` APIs are not considered done.  So feel free to open an issue in [their repo](https://github.com/dotnet/command-line-api).  What you're suggesting makes sense to me.

**Comment by SteveL-MSFT** [comment]
This has been replaced by https://github.com/PowerShell/PowerShell/issues/13071
